### ISSUE TITLE
SDL2, wayland backend: Wayland scaling makes cursor smaller but it should be bigger

### ISSUE THREAD

**Issue by ernstp** [libsdl-org/SDL#9250]
Testing with Plasma 6.0.1, Wayland, SDL 2.30.1, Factorio with `SDL_DYNAMIC_API=/usr/lib64/libSDL2-2.0.so.0 SDL_VIDEODRIVER=wayland %command%`

I have monitor scale set to 150%. When I start Factorio, I get an incredibly tiny mouse cursor. It looks like the scaling is inverted, I'd need the cursor to cover more physical pixels.
Doesn't happen if I set SDL_VIDEODRIVER=x11

**Comment by Kontrabant** [comment]
Do you have the xdg-desktop-portal-gtk package installed?

It seems that KDE 6 no longer sets the `XCURSOR_SIZE` envvar, and KDE based distros may not have the xdg-desktop-portal-gtk package installed by default, leaving no way for SDL to query the cursor size.

The long-term solution is to add support for the cursor-shape-v1 protocol, now that a compositor actually supports it, but in the meantime, try installing the xdg-desktop-portal-gtk package if it isn't already. That will allow SDL to query the cursor size, even when using KDE.

**Comment by ernstp** [comment]
I do have xdg-desktop-portal-gtk installed already, v1.15.1 .
Also setting XCURSOR_SIZE to different values didn't seem to have any effect...

Perhaps the query through the portal works but the answer is not what it expects?

**Comment by Kontrabant** [comment]
Thanks for the additional info. I'm grabbing a KDE Neon ISO and will look into this.

**Comment by Kontrabant** [comment]
Hmm, I grabbed the Factorio demo and the tiny cursor issue is present on GNOME too.

**Comment by ernstp** [comment]
Ok thanks for looking into this! I was sure I had used Factorio to test and debug libSDL before with the dynapi.

**Comment by ernstp** [comment]
Hmm if I set XCURSOR_THEME=Adwaita (that's a Gnome theme) it seems to work quite ok actually... then XCURSOR_SIZE=128 has an effect. Otherwise it has no effect.
Probably not an SDL issue. Just leaving it here for reference.

**Comment by ernstp** [comment]
So Plasma 5 is exporting the XCURSOR_THEME variable but not Plasma 6, so that's the cause of why I'm seeing this now.
https://invent.kde.org/plasma/kwin/-/merge_requests/3936

However the dbus call should still work?
dbus-send --session --print-reply --dest=org.freedesktop.portal.Desktop /org/freedesktop/portal/desktop org.freedesktop.portal.Settings.Read string:org.gnome.desktop.interface string:cursor-theme
method return time=1710252994.682134 sender=:1.5 -> destination=:1.182 serial=2727 reply_serial=2
   variant       variant          string "Adwaita"

Then again, I'm not sure exactly what version of libSDL2 Factorio is using, so this should be reproduced with vanilla libSDL2 from git first, or something like that...

**Comment by Kontrabant** [comment]
There are definitely improvements to be made to the Wayland cursor handling code, such as closest size matches for cursor selection and supporting the new cursor protocol, both of which I plan on working on shortly, but the root cause here is the SDL build they statically linked into the game. It's either too old to support the dbus settings, which weren't added all that many versions ago, or they built it without dbus support, which means that any dbus related code is disabled, which I'm guessing is the case, as searching for dbus functions in the executable turned up nothing either. Either way, they have to fix it, because with dynapi disabled, it's out of our hands.

**Comment by ernstp** [comment]
> or they built it without dbus support

Ooh, I see, that's probably it...

So workaround for Plasma 6 is XCURSOR_THEME=your_theme_name %command% anyway then.

**Comment by Kontrabant** [comment]
The cursor shape protocol support has been backported to SDL2, so we've done everything we can do here.

**Comment by raiguard** [comment]
I just came across this thread. Building SDL without dbus support was a simple oversight and I will fix that. I confirmed that building with dbus support fixes the cursor issue on GNOME.

However, SDL2 2.30.9 does not appear to contain `cursor-shape-v1` code. The code exists on the `SDL2` branch, but in the `2.30.9` release tag, it is absent. This means that cursors will continue to be broken on Plasma. Was this intentional?
